opendatafabric/src/dtos/dtos_extra.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::borrow::Cow;
+use std::fmt::Display;
+
+use super::DatasetVocabulary;
+use crate::{
+    ExecuteQueryResponseInternalError,
+    ExecuteQueryResponseInvalidQuery,
+    ReadStepCsv,
+    SetVocab,
+    SqlQueryStep,
+    TransformSql,
+};
+
+////////////////////////////////////////////////////////////////////////////////
+// DatasetVocabulary
+////////////////////////////////////////////////////////////////////////////////
+
+impl DatasetVocabulary {
+    pub fn into_resolved(self) -> DatasetVocabularyResolvedOwned {
+        self.into()
+    }
+}
+
+impl Default for DatasetVocabulary {
+    fn default() -> Self {
+        DatasetVocabulary {
+            offset_column: None,
+            system_time_column: None,
+            event_time_column: None,
+        }
+    }
+}
+
+impl From<SetVocab> for DatasetVocabulary {
+    fn from(v: SetVocab) -> Self {
+        Self {
+            offset_column: v.offset_column,
+            system_time_column: v.system_time_column,
+            event_time_column: v.event_time_column,
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+pub struct DatasetVocabularyResolved<'a> {
+    /// Name of the system time column.
+    pub system_time_column: Cow<'a, str>,
+    /// Name of the event time column.
+    pub event_time_column: Cow<'a, str>,
+    /// Name of the offset column.
+    pub offset_column: Cow<'a, str>,
+}
+
+pub type DatasetVocabularyResolvedOwned = DatasetVocabularyResolved<'static>;
```

Comment:
This is a really cool trick I picked up from `datafusion` - it allows to have a single type that can contain both borrowed references with limited lifetime and owned objects (with `'static`) lifetime.